flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/data/StructRowData.java

```diff
@@ -0,0 +1,340 @@
+        collectionToArrayData(mapType.keyType(), Lists.transform(entries, Map.Entry::getKey));
+
+    ArrayData valueArray =
+        collectionToArrayData(mapType.valueType(), Lists.transform(entries, Map.Entry::getValue));
+
+    ArrayData.ElementGetter keyGetter =
+        ArrayData.createElementGetter(FlinkSchemaUtil.convert(mapType.keyType()));
+    ArrayData.ElementGetter valueGetter =
+        ArrayData.createElementGetter(FlinkSchemaUtil.convert(mapType.valueType()));
+
+    int length = keyArray.size();
+    Map<Object, Object> result = Maps.newHashMap();
+    for (int pos = 0; pos < length; pos++) {
+      final Object keyValue = keyGetter.getElementOrNull(keyArray, pos);
+      final Object valueValue = valueGetter.getElementOrNull(valueArray, pos);
+
+      result.put(keyValue, valueValue);
+    }
+    return new GenericMapData(result);
+  }
+
+  private ArrayData collectionToArrayData(Type elementType, Collection<?> values) {
+    switch (elementType.typeId()) {
+      case BOOLEAN:
+      case INTEGER:
+      case DATE:
+      case TIME:
+      case LONG:
+      case TIMESTAMP:
+      case FLOAT:
+      case DOUBLE:
+        return fillArray(values, array -> (pos, value) -> array[pos] = value);
+      case STRING:
+        return fillArray(
+            values,
+            array ->
+                (BiConsumer<Integer, CharSequence>)
+                    (pos, seq) -> array[pos] = StringData.fromString(seq.toString()));
+      case FIXED:
+      case BINARY:
+        return fillArray(
+            values,
+            array ->
+                (BiConsumer<Integer, ByteBuffer>)
+                    (pos, buf) -> array[pos] = ByteBuffers.toByteArray(buf));
+      case DECIMAL:
+        return fillArray(
+            values, array -> (BiConsumer<Integer, BigDecimal>) (pos, dec) -> array[pos] = dec);
+      case STRUCT:
+        return fillArray(
+            values,
+            array ->
+                (BiConsumer<Integer, StructLike>)
+                    (pos, tuple) ->
+                        array[pos] = new StructRowData(elementType.asStructType(), tuple));
+      case LIST:
+        return fillArray(
+            values,
+            array ->
+                (BiConsumer<Integer, Collection<?>>)
+                    (pos, list) ->
+                        array[pos] =
+                            collectionToArrayData(elementType.asListType().elementType(), list));
+      case MAP:
+        return fillArray(
+            values,
+            array ->
+                (BiConsumer<Integer, Map<?, ?>>)
+                    (pos, map) -> array[pos] = new GenericMapData(map));
+      default:
+        throw new UnsupportedOperationException("Unsupported array element type: " + elementType);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T> GenericArrayData fillArray(
+      Collection<?> values, Function<Object[], BiConsumer<Integer, T>> makeSetter) {
+    Object[] array = new Object[values.size()];
```

Comment:
is this correct? should the array corresponds to the element type of the Iceberg list? 

e.g., FlinkOrcReaders
```
    @Override
    public ArrayData nonNullRead(ColumnVector vector, int row) {
      ListColumnVector listVector = (ListColumnVector) vector;
      int offset = (int) listVector.offsets[row];
      int length = (int) listVector.lengths[row];
      List<T> elements = Lists.newArrayListWithExpectedSize(length);
      for (int c = 0; c < length; ++c) {
        elements.add(elementReader.read(listVector.child, offset + c));
      }
      return new GenericArrayData(elements.toArray());
    }
```

or TestRowProjection
```
  @Test
  public void testListProjection() throws IOException {
    Schema writeSchema =
        new Schema(
            Types.NestedField.required(0, "id", Types.LongType.get()),
            Types.NestedField.optional(
                10, "values", Types.ListType.ofOptional(11, Types.LongType.get())));

    GenericArrayData values = new GenericArrayData(new Long[] {56L, 57L, 58L});
```